### PR TITLE
Allow directory entries in EXTRA_PATH_METADATA

### PR DESCRIPTION
**Pull request by andrew-vant**
This lets the user apply arbitrary metadata to everything in a directory. My specific use case is changing the stylesheet and template, but it also provides workarounds for cases when DEFAULT_METADATA isn't enough; e.g. applying the [default draft recipe](http://docs.getpelican.com/en/3.6.3/content.html#publishing-drafts) from the docs may not work as expected because [pages and articles accept different status types](https://github.com/getpelican/pelican/issues/1620#issuecomment-73736406).

### CODE REVIEW

**Comment by justinmayer** [comment]
Any comments on this PR from @getpelican/reviewers would be much appreciated. What say you?

**Comment by ingwinlu** [comment]
like the idea, but the usual suspects i.e. test cases and documentation is missing.

**Comment by andrew-vant** [comment]
I did update the documentation -- or at least I thought I did. Is there another piece I should have included?

Acknowledged on test cases. I'll fix that.

**Comment by ingwinlu** [comment]
@andrew-vant nah it's okay, thought we had more on that...

Since you worked on that part of pelican, would you say it is possible to consolidate all the Metadata settings(`FILENAME_METADATA`, `PATH_METADATA` and `EXTRA_PATH_METADATA`) into one?

**Comment by andrew-vant** [comment]
I don't think so, or at least not easily. Some context: When I first started messing with EXTRA_PATH_METADATA, I wondered why it doesn't take regexes. On #pelican, Avaris told me part of the reason is that there's no good deterministic way to handle multiple conflicting matches. I think they were right.

PATH_METADATA and FILENAME_METADATA both rely on regex to do their magic, and EXTRA_PATH_METADATA relies on regexlessness to be deterministic. You could maybe eliminate FILENAME_METADATA on the grounds that everything it can do can already be done by PATH_METADATA, but I don't know if it would gain much.

**Comment by justinmayer** [comment]
Hey Andrew. Many thanks for your contribution. Would you mind adding some tests so we can get it merged?

**Comment by andrew-vant** [comment]
Oops. Back when I opened this, I started using my own branch and forgot I still had the PR open. I'll see what I can do about tests. Thanks for the reminder.

**Comment by andrew-vant** [comment]
Okay, tests added. One test fails on my local machine, but it also failed on the commit I branched from, so I'm guessing it's not my problem.

It's been a long time since I opened this PR. I can rebase the branch on master if you want me to.

**Comment by andrew-vant** [comment]
Ping? Not sure if I need to do anything with this, but I noticed it's still open a couple months later.

**Comment by justinmayer** [comment]
Any last comments from @getpelican/reviewers?

File: pelican/readers.py [hunk 1]
```diff
@@ -630,8 +630,20 @@ def path_metadata(full_path, source_path, settings=None):
...
+        # most specific entry wins.
+
+        paths = []
+        path = source_path
+        while path:
+            paths.append(path)
+            path = os.path.split(path)[0]
+        epm = settings.get('EXTRA_PATH_METADATA', {})
+        for path in reversed(paths):
+            metadata.update(epm.get(path, {}))
```
Comment (oulenz) [inline pelican/readers.py]: I'm wondering whether it is really necessary to break down the `source_path` into its ancestors like that, and whether it wouldn't be simpler to try to match the `epm` paths:

```
epm = [(p, m) for p, m in epm.items() if p == source_path or source_path.startswith(os.path.join(p, '')))
for _, m in sorted(epm):
    metadata.update(m)
```
In particular, this would automatically skip all comparison if `epm` is empty. (This version also tolerates trailing slashes, so we don't need to warn people against those.)

If you prefer your own version, could you at least add a check whether `epm` isn't empty?

**Comment by oulenz** [comment]
I've added a stylistic comment, and maybe the commits should be squashed. Otherwise, this looks good to me.

File: pelican/readers.py (hunk 1 above)
Comment (avaris) [reply pelican/readers.py]: I think it was done this way to guarantee "if multiple EPM entries apply conflicting metadata to the same file, the most specific entry wins."

Comment (andrew-vant) [reply pelican/readers.py]: I think avaris is right, but it's been a while so I don't remember for sure.

**Comment by oulenz** [comment]
Of course, but that's not the point I was trying to make, and my version
should also let more specific paths overrule others (that's what `sorted`
ensures).

Den man. 25. feb. 2019, 23.21 skrev Andrew Vant <notifications@github.com>:

File: pelican/readers.py (hunk 1 above)
Comment (oulenz) [reply pelican/readers.py]: Of course, but that's not the point I was trying to make, and my version should also let more specific paths overrule others (that's what `sorted` ensures). The following is probably clearer:

```
for p, m in sorted(epm.items()):
    if p == source_path or source_path.startswith(os.path.join(p, ''))):
        metadata.update(m)
```

Comment (andrew-vant) [reply pelican/readers.py]: Hrm. That is clearer, and easier to follow than my version. What does os.path.join do with a blank string, exactly? I'm guessing it adds a slash if one isn't present. I worry that .startswith will produce false positives, but I'm having trouble coming up with a  path that would cause it.

I'll fiddle with it later and maybe update the PR.

Comment (oulenz) [reply pelican/readers.py]: - yes, it adds a slash if one isn't present. It's a bit hacky, admittedly.
- I don't see how `startswith` could produce false positives with the final slash.

The main thing that got me thinking was that your version generates all the ancestors of `source_path` even if `epm` is empty, which is probably the case for 99% of installations. As long as you address that, I'm happy. :-)

Comment (avaris) [reply pelican/readers.py]: Is slash necessary? Can't think of a reason why `source_path.startswith(p)` isn't enough.

Comment (oulenz) [reply pelican/readers.py]: Potentially you could have articles `foo/bar.md` and `foobar/baz.md`, then if you key extra metadata with `foo`, you only want it to affect the former, not the latter.

Comment (avaris) [reply pelican/readers.py]: Oh, right :).

**Comment by andrew-vant** [comment]
Updated to use Oulenz's suggestion, and I rebased/squashed the commits too. (I still have the old ones if needed)

File: pelican/tests/test_readers.py [hunk 2]
```diff
@@ -273,6 +273,37 @@ def test_article_extra_path_metadata_dont_overwrite(self):
...
+
+        # Make sure vars aren't getting "inherited" by mistake...
+        path = "article.rst"
+        page = self.read_file(path=path, EXTRA_PATH_METADATA=epm)
+        for k in expected_metadata.keys():
+            self.assertNotIn(k, page.metadata)
+
+        # Same, but for edge cases where one file's name is a prefix of
+        # another.
+        page = self.read_file(path=path, EXTRA_PATH_METADATA=epm)
```
Comment (oulenz) [inline pelican/tests/test_readers.py]: I think you want to test this with `path = "TestCategory/article_without_category.rst"` as defined at the beginning of this test, but you've redefined  `path` to `"article.rst"` in the previous block.

Comment (andrew-vant) [reply pelican/tests/test_readers.py]: Oops, yes, thanks. Fixed.

**Comment by justinmayer** [comment]
@avaris: Everything look okay to you here?

**Comment by justinmayer** [comment]
Many thanks to Andrew for the enhancement and to @ingwinlu, @avaris, and @oulenz for reviewing.
